File: arrays_and_strings/meeting_rooms.py
# 253. Meeting Rooms II
#
# Time complexity: O(NlogN)
# Space complexity O(N)
#
# Intuition:
#
# The number of rooms required increases when meetings overlap.
# To check if two meetings overlap we can use the end and start time.
# We will iterate over the intervals and check on each meeting if there are meetings going on.
# If there are meetings, we'll check and remove the ones that have already finished.
# Once we remove all the finished meetings, then we proceed to schedule the new meeting.
#
# The number of rooms required is the max number reached by our schedule. i.e. the max number of meetings that overlapped during the day
# The ideal data structure to store the meetings is a HEAP, because we will want to remove the meetings with the LOWEST endtime time first.

from heapq import *

def rooms_required(intervals):

    # sort intervals by start time
    intervals.sort(key= lambda x: x[0])

    # keep meeting ending times in a heap
    schedule = []

    # keep track of the max size o the heap
    rooms = 0

    # check all meetings for the day
    for meeting in intervals:
        # remove meetings that have ended before start time of this meeting (if any)
        while schedule and schedule[0] <= meeting[0]:
            heappop(schedule)

        # schedule the meeting and keep ending time 
        heappush(schedule, meeting[1])

        rooms = max(rooms, len(schedule))

    return rooms

intervals_1 = [[0,30],[5,10],[15,20]]
intervals_2 = [[7,10],[2,4]]

assert rooms_required(intervals_1) == 2
assert rooms_required(intervals_2) == 1

print("all tests passed.")

File: arrays_and_strings/meeting_rooms.php
<?php

// Meeting rooms
echo "[ Meeting rooms ] \n";

class Meeting 
{
    public function __construct($start, $end)
    {
        $this->start = $start;
        $this->end = $end;
    }
}

class Solution
{
    function meeting_rooms($meetings) {
        // max number of rooms 
        $max_rooms = 0;

        // sort meetings by start time
        usort($meetings, fn($a, $b) => $a->start - $b->start);

        // create heap
        $schedule = new SplMinHeap();
        
        // traverse meetings data
        foreach ($meetings as $meeting) {
            while ($schedule->count() > 1 && $schedule->top() <= $meeting->start) {
                $schedule->extract();
            }

            $schedule->insert($meeting->end);
            $max_rooms = max($max_rooms, $schedule->count());
        }

        return $max_rooms;
    }
}

// tests
$meetings_1 = array(
    new Meeting(8,9),
    new Meeting(1,5),
    new Meeting(2,6),
    new Meeting(11,12),
    new Meeting(4,5)
);

$meetings_2 = array(
    new Meeting(0,30),
    new Meeting(5,10),
    new Meeting(10,20)
);

$solution = new Solution();
$result_1 = $solution->meeting_rooms($meetings_1);
$result_2 = $solution->meeting_rooms($meetings_2);

echo "\e[1;37;42mTest 1: " . $result_1 . "\n\e[0m\n";
echo "\e[1;37;42mTest 2: " . $result_2 . "\n\e[0m\n";

?>